src/main/java/mate.academy/lib/Injector.java

```diff
@@ -1,13 +1,79 @@
+import mate.academy.service.impl.FileReaderServiceImpl;
+import mate.academy.service.impl.ProductParserImpl;
+import mate.academy.service.impl.ProductServiceImpl;
+
 public class Injector {
     private static final Injector injector = new Injector();
+    private static final Map<Class<?>, Class<?>> interfaceImplementations =
+            Map.of(ProductService.class, ProductServiceImpl.class,
+            ProductParser.class, ProductParserImpl.class,
+            FileReaderService.class, FileReaderServiceImpl.class);
+    private Map<Class<?>, Object> instances = new HashMap<>();
 
     public static Injector getInjector() {
         return injector;
     }
 
     public Object getInstance(Class<?> interfaceClazz) {
-        return null;
+        Class<?> clazz = findImplementation(interfaceClazz);
+        if (!clazz.isAnnotationPresent(Component.class)) {
+            throw new RuntimeException("Injection failed, missing "
+                    + "@Component annotation on the class "
+                    + interfaceClazz.getName());
+        }
+        Field[] fields = clazz.getDeclaredFields();
+        Object classImplementationInstance = null;
+        for (Field field : fields) {
+            if (field.isAnnotationPresent(Inject.class)) {
+                Object fieldInstance = getInstance(field.getType());
+                classImplementationInstance = createNewInstance(clazz);
+                try {
+                    field.setAccessible(true);
+                    field.set(classImplementationInstance, fieldInstance);
+                } catch (IllegalAccessException e) {
+                    throw new RuntimeException("Can't initialize a field value. Class: "
+                            + clazz.getName()
+                            + ". Field: " + field.getName(), e);
+                }
+            }
+        }
+        if (classImplementationInstance == null) {
+            classImplementationInstance = createNewInstance(clazz);
+        }
+        return classImplementationInstance;
+    }
+
+    private Object createNewInstance(Class<?> clazz) {
+        if (instances.containsKey(clazz)) {
+            return instances.get(clazz);
+        }
+        try {
+            Constructor<?> constructor = clazz.getConstructor();
+            Object instance = constructor.newInstance();
+            instances.put(clazz, instance);
+            return instance;
+        } catch (NoSuchMethodException e) {
+            throw new RuntimeException("Cannot find a constructor at " + clazz.getName(), e);
+        } catch (InvocationTargetException e) {
+            throw new RuntimeException("Cannot call a constructor at " + clazz.getName(), e);
+        } catch (InstantiationException e) {
+            throw new RuntimeException("Cannot instantiate " + clazz.getName(), e);
+        } catch (IllegalAccessException e) {
+            throw new RuntimeException("Cannot create a new instance of " + clazz.getName(), e);
+        }
+    }
```

Comment:
<img width="798" alt="image" src="https://github.com/mate-academy/jv-dependency-injection/assets/81781479/913eef2d-4317-4567-8ba4-b45789edcd53">


not fixed